// cdktf/lib/importable-resource.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImportableResource = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const terraform_element_1 = require("./terraform-element");
const validate_terraform_version_1 = require("./validations/validate-terraform-version");
/**
 * Class used to represent an importable resource.
 */
class ImportableResource extends terraform_element_1.TerraformElement {
    constructor(scope, name, config) {
        super(scope, name, config.terraformResourceType);
        this.config = config;
        this.node.addValidation(new validate_terraform_version_1.ValidateTerraformVersion(">=1.5", `Import blocks are only supported for Terraform >=1.5. Please upgrade your Terraform version.`));
    }
    toHclTerraform() {
        const expectedResourceAddress = `${this.config.terraformResourceType}.${this.friendlyUniqueId}`;
        return {
            import: [
                {
                    to: {
                        value: expectedResourceAddress,
                        type: "simple",
                        storageClassType: "reference",
                    },
                    id: {
                        value: this.config.importId,
                        type: "simple",
                        storageClassType: "string",
                    },
                    provider: this.config.provider
                        ? {
                            value: this.config.provider.fqn,
                            type: "simple",
                            storageClassType: "reference",
                        }
                        : undefined,
                },
            ],
        };
    }
    toTerraform() {
        const expectedResourceAddress = `${this.config.terraformResourceType}.${this.friendlyUniqueId}`;
        return {
            import: [
                {
                    to: expectedResourceAddress,
                    id: this.config.importId,
                    provider: this.config.provider ? this.config.provider.fqn : undefined,
                },
            ],
        };
    }
    toMetadata() {
        return {
            importsGeneratingConfiguration: [this.friendlyUniqueId],
        };
    }
}
exports.ImportableResource = ImportableResource;
_a = JSII_RTTI_SYMBOL_1;
ImportableResource[_a] = { fqn: "cdktf.ImportableResource", version: "0.21.0" };
