// cdktf/lib/runtime.js
"use strict";
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
// ----------------------------------------------------------------------
// PROPERTY MAPPERS
//
// These are used while converting generated classes/property bags to Terraform JSON objects
//
// We use identity mappers for the primitive types. These don't do anything but are there to make the code
// generation work out nicely (so the code generator doesn't need to emit different code for primitive
Object.defineProperty(exports, "__esModule", { value: true });
exports.isComplexElement = exports.numberToHclTerraform = exports.anyToHclTerraform = exports.booleanToHclTerraform = exports.stringToHclTerraform = exports.numberToTerraform = exports.anyToTerraform = exports.booleanToTerraform = exports.stringToTerraform = void 0;
exports.listMapper = listMapper;
exports.listMapperHcl = listMapperHcl;
exports.hashMapper = hashMapper;
exports.hashMapperHcl = hashMapperHcl;
exports.canInspect = canInspect;
const encoding_1 = require("./tokens/private/encoding");
Object.defineProperty(exports, "isComplexElement", { enumerable: true, get: function () { return encoding_1.isComplexElement; } });
const tokens_1 = require("./tokens");
const terraform_dynamic_block_1 = require("./terraform-dynamic-block");
const terraform_dynamic_expression_1 = require("./terraform-dynamic-expression");
// eslint-disable-next-line jsdoc/require-jsdoc
function identity(x) {
    return x;
}
exports.stringToTerraform = identity;
exports.booleanToTerraform = identity;
exports.anyToTerraform = identity;
exports.numberToTerraform = identity;
exports.stringToHclTerraform = identity;
exports.booleanToHclTerraform = identity;
exports.anyToHclTerraform = identity;
exports.numberToHclTerraform = identity;
/**
 * @param isBlockType blocks and lists of objects need to be handled different in Terraform
 *                    but are represented exactly the same in generated bindings
 *                    currently this is used to resolve TerraformDynamicBlocks if not within
 *                    a block type (will be resolved to a for expression for those instead of
 *                    a dynamic block)
 */
function listMapper(elementMapper, isBlockType) {
    return (x) => {
        if (!canInspect(x)) {
            return x;
        }
        // replace dynamic expressions for block types so they can be detected and replaced properly by processDynamicAttributes()
        // which also relocates them to dynamic.attributeName (something we can't do when resolving a value, put it somewhere else)
        // if isBlockType is false, the TerraformDynamicExpression that is present will be resolved (it implements IResolvable) to a
        // for expression directly (which can be used e.g. within Terraform functions or for attributes that are not of a block type)
        if (terraform_dynamic_expression_1.TerraformDynamicExpression.isTerraformDynamicExpression(x) &&
            isBlockType) {
            return terraform_dynamic_block_1.TerraformDynamicBlock.fromDynamicExpression(x);
        }
        if (!Array.isArray(x)) {
            return x;
        }
        return x.map(elementMapper);
    };
}
/**
 * @param isBlockType blocks and lists of objects need to be handled different in Terraform
 *                    but are represented exactly the same in generated bindings
 *                    currently this is used to resolve TerraformDynamicBlocks if not within
 *                    a block type (will be resolved to a for expression for those instead of
 *                    a dynamic block)
 */
function listMapperHcl(elementMapper, isBlockType) {
    return (x) => {
        if (!canInspect(x)) {
            return x;
        }
        // replace dynamic expressions for block types so they can be detected and replaced properly by processDynamicAttributes()
        // which also relocates them to dynamic.attributeName (something we can't do when resolving a value, put it somewhere else)
        // if isBlockType is false, the TerraformDynamicExpression that is present will be resolved (it implements IResolvable) to a
        // for expression directly (which can be used e.g. within Terraform functions or for attributes that are not of a block type)
        if (terraform_dynamic_expression_1.TerraformDynamicExpression.isTerraformDynamicExpression(x) &&
            isBlockType) {
            return terraform_dynamic_block_1.TerraformDynamicBlock.fromDynamicExpression(x);
        }
        if (!Array.isArray(x)) {
            return x;
        }
        return x.map(elementMapper);
    };
}
// eslint-disable-next-line jsdoc/require-jsdoc
function hashMapper(elementMapper) {
    return (x) => {
        if (!canInspect(x)) {
            return x;
        }
        // Check if it's a token object
        if ((0, encoding_1.containsComplexElement)(x)) {
            return x;
        }
        if (tokens_1.Tokenization.isResolvable(x)) {
            return x;
        }
        // We can't treat strings as hashes (likely a token or a misconfiguration)
        if (typeof x === "string") {
            return x;
        }
        if ((0, encoding_1.containsMapToken)(x)) {
            return x;
        }
        const ret = {};
        Object.keys(x).forEach((key) => {
            ret[key] = elementMapper(x[key]);
        });
        return ret;
    };
}
/**
 *
 */
function hashMapperHcl(elementMapper) {
    return (x) => {
        if (!canInspect(x)) {
            return x;
        }
        // Check if it's a token object
        if ((0, encoding_1.containsComplexElement)(x)) {
            return x;
        }
        if (tokens_1.Tokenization.isResolvable(x)) {
            return x;
        }
        // We can't treat strings as hashes (likely a token or a misconfiguration)
        if (typeof x === "string") {
            return x;
        }
        if ((0, encoding_1.containsMapToken)(x)) {
            return x;
        }
        const ret = {};
        Object.keys(x).forEach((key) => {
            ret[key] = elementMapper(x[key]);
        });
        return ret;
    };
}
/**
 * Return whether this object can be validated at all
 *
 * True unless it's undefined
 */
function canInspect(x) {
    // Note: using weak equality on purpose, we also want to catch undefined
    return x != null;
}
