// cdktf/lib/terraform-dynamic-expression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraformDynamicExpression = void 0;
const _1 = require(".");
const tokens_1 = require("./tokens");
const stack_trace_1 = require("./tokens/private/stack-trace");
const DYNAMIC_EXPRESSION_SYMBOL = Symbol.for("cdktf/TerraformDynamicExpression");
/**
 * A TerraformDynamicExpression is returned by TerraformIterator.dynamic
 * and can be passed to list attributes of e.g. TerraformResources
 * As it resolves to a for expression, it can be used directly on attributes of
 * a list(object) type (or similar) and in Terraform functions that take a list.
 * When used in attributes that are of a block type which need to be modeled as
 * a dynamic block, the listMapper for that attribute will change this to a
 * TerraformDynamicBlock, which in turn will be properly handled by the
 * processDynamicAttributes() utility function.
 */
class TerraformDynamicExpression {
    constructor(args) {
        Object.defineProperty(this, DYNAMIC_EXPRESSION_SYMBOL, { value: true });
        this.creationStack = (0, stack_trace_1.captureStackTrace)();
        this.iterator = args.iterator;
        this.content = args.content;
    }
    /**
     * resolves this as a Terraform for expression.
     * This allows using iterator.dynamic() on attributes that are not blocks but
     * lists of objects (which is not visible to the user when working with provider
     * mappings as it is abstracted away)
     */
    resolve() {
        return tokens_1.Lazy.anyValue({
            produce: (context) => {
                // resolve nested Lazy values (within this.content) for a for expression
                context.iteratorContext = "FOR_EXPRESSION";
                // returns [ for key, value in forEachExpr : content ]
                // context.resolve is required for the iteratorContext to be correctly passed
                // to Lazy values within this.content
                return context.resolve((0, _1.forExpression)(this.iterator._getForEachExpression(), this.content));
            },
        });
    }
    toString() {
        return tokens_1.Token.asString(this);
    }
    /**
     * Turn this Dynamic Expression into JSON
     *
     * Called automatically when JSON.stringify() is called on this.
     */
    toJSON() {
        return "<unresolved-terraform-dynamic-expression>";
    }
    static isTerraformDynamicExpression(x) {
        return (x !== null && typeof x === "object" && DYNAMIC_EXPRESSION_SYMBOL in x);
    }
}
exports.TerraformDynamicExpression = TerraformDynamicExpression;
