// cdktf/lib/terraform-functions.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fn = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
const _1 = require(".");
const helpers_1 = require("./functions/helpers");
const terraform_functions_generated_1 = require("./functions/terraform-functions.generated");
// eslint-disable-next-line jsdoc/require-jsdoc
class Fn extends terraform_functions_generated_1.FnGenerated {
    /**
     * {@link /terraform/docs/language/functions/bcrypt.html bcrypt} computes a hash of the given string using the Blowfish cipher, returning a string in [the _Modular Crypt Format_](https://passlib.readthedocs.io/en/stable/modular_crypt_format.html) usually expected in the shadow password file on many Unix systems.
     * @param {string} str
     * @param {number} [cost]
     */
    static bcrypt(str, cost) {
        // overwritten because bcrypt() only supports a single variadic parameter and not multiple
        // There is currently nothing in the schema that represents this runtime check:
        // https://github.com/hashicorp/terraform/blob/6ab3faf5f65a90ae1e5bd0625fa9e83c0b34c5e1/internal/lang/funcs/crypto.go#L115-L117
        return Fn._bcrypt(str, cost ? [cost] : []);
    }
    /**
     * {@link https://developer.hashicorp.com/terraform/language/expressions/conditionals} A conditional expression uses the value of a boolean expression to select one of two values.
     * @param {Expression} condition
     * @param {Expression} trueValue
     * @param {Expression} falseValue
     */
    static conditional(condition, trueValue, falseValue) {
        return (0, _1.conditional)(condition, trueValue, falseValue);
    }
    /**
     * {@link /terraform/docs/language/functions/lookup.html lookup} retrieves the value of a single element from a map, given its key. If the given key does not exist, the given default value is returned instead.
     * @param {any} inputMap
     * @param {string} key
     * @param {any} [defaultValue]
     */
    static lookup(inputMap, key, defaultValue) {
        // overwritten because lookup() uses a variadic argument for its optional defaultValue
        if (defaultValue)
            return Fn._lookup(inputMap, key, [defaultValue]);
        return (0, helpers_1.asAny)((0, _1.propertyAccess)(inputMap, [key])); // -> renders inputMap[key] (which is recommended if no default value is given)
    }
    /**
     * returns a property access expression that accesses the property at the given path in the given inputMap.
     * For example lookupNested(x, ["a", "b", "c"]) will return a Terraform expression like x["a"]["b"]["c"]
     * @param {any} inputMap
     * @param {Array<any>} path
     */
    static lookupNested(inputMap, path) {
        return (0, helpers_1.asAny)((0, _1.propertyAccess)(inputMap, path));
    }
    /**
     * {@link /terraform/docs/language/functions/join.html join} produces a string by concatenating together all elements of a given list of strings with the given delimiter.
     * @param {string} separator
     * @param {Array<string>} list
     */
    static join(separator, list) {
        // overwritten because join() supports passing multiple lists to it e.g. join(sep, listA, listB)
        // which we can't model as JSII does not support variadic arguments
        return Fn._join(separator, [list]);
    }
    /**
     * {@link /terraform/docs/language/functions/range.html range} generates a list of numbers using a start value, a limit value, and a step value.
     * @param {number} start
     * @param {number} limit
     * @param {number} [step=1]
     */
    static range(start, limit, step = 1) {
        // overwritten because the step param in range() is optional which
        // the Terraform functions schema represents as a single variadic argument
        return Fn._range([start, limit, step]);
    }
    /**
     * Use this function to wrap a string and escape it properly for the use in Terraform
     * This is only needed in certain scenarios (e.g., if you have unescaped double quotes in the string)
     * @param {String} str
     */
    static rawString(str) {
        return _1.Token.asString((0, _1.rawString)(str));
    }
}
exports.Fn = Fn;
_a = JSII_RTTI_SYMBOL_1;
Fn[_a] = { fqn: "cdktf.Fn", version: "0.21.0" };
