// cdktf/lib/terraform-operators.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Op = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */
const tfExpression_1 = require("./tfExpression");
const tokens_1 = require("./tokens");
/**
 * This class contains static functions for all arithmetical and logical operators
 * in the Terraform configuration language.
 */
class Op {
    /**
     * Renders !expression
     */
    static not(expression) {
        return new tfExpression_1.OperatorExpression("!", expression);
    }
    /**
     * Renders -expression
     */
    static negate(expression) {
        return tokens_1.Token.asNumber(new tfExpression_1.OperatorExpression("-", expression));
    }
    /**
     * Renders left * right
     */
    static mul(left, right) {
        return tokens_1.Token.asNumber(new tfExpression_1.OperatorExpression("*", left, right));
    }
    /**
     * Renders left / right
     */
    static div(left, right) {
        return tokens_1.Token.asNumber(new tfExpression_1.OperatorExpression("/", left, right));
    }
    /**
     * Renders left % right
     */
    static mod(left, right) {
        return tokens_1.Token.asNumber(new tfExpression_1.OperatorExpression("%", left, right));
    }
    /**
     * Renders left + right
     */
    static add(left, right) {
        return tokens_1.Token.asNumber(new tfExpression_1.OperatorExpression("+", left, right));
    }
    /**
     * Renders left - right
     */
    static sub(left, right) {
        return tokens_1.Token.asNumber(new tfExpression_1.OperatorExpression("-", left, right));
    }
    /**
     * Renders left > right
     */
    static gt(left, right) {
        return new tfExpression_1.OperatorExpression(">", left, right);
    }
    /**
     * Renders left >= right
     */
    static gte(left, right) {
        return new tfExpression_1.OperatorExpression(">=", left, right);
    }
    /**
     * Renders left < right
     */
    static lt(left, right) {
        return new tfExpression_1.OperatorExpression("<", left, right);
    }
    /**
     * Renders left <= right
     */
    static lte(left, right) {
        return new tfExpression_1.OperatorExpression("<=", left, right);
    }
    /**
     * Renders left == right
     */
    static eq(left, right) {
        return new tfExpression_1.OperatorExpression("==", left, right);
    }
    /**
     * Renders left != right
     */
    static neq(left, right) {
        return new tfExpression_1.OperatorExpression("!=", left, right);
    }
    /**
     * Renders left && right
     */
    static and(left, right) {
        return new tfExpression_1.OperatorExpression("&&", left, right);
    }
    /**
     * Renders left || right
     */
    static or(left, right) {
        return new tfExpression_1.OperatorExpression("||", left, right);
    }
}
exports.Op = Op;
_a = JSII_RTTI_SYMBOL_1;
Op[_a] = { fqn: "cdktf.Op", version: "0.21.0" };
