// cdktf/lib/terraform-count.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraformCount = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
const tfExpression_1 = require("./tfExpression");
const tokens_1 = require("./tokens");
const COUNT_SYMBOL = Symbol.for("cdktf/TerraformCount");
/**
 * Iterator for the Terraform count property
 */
class TerraformCount {
    static of(count) {
        return new TerraformCount(count);
    }
    static isTerraformCount(x) {
        return x !== null && typeof x === "object" && COUNT_SYMBOL in x;
    }
    constructor(count) {
        this.count = count;
        Object.defineProperty(this, COUNT_SYMBOL, { value: true });
    }
    get index() {
        return tokens_1.Token.asNumber((0, tfExpression_1.ref)("count.index"));
    }
    toTerraform() {
        return this.count;
    }
    toString() {
        return this.count.toString();
    }
}
exports.TerraformCount = TerraformCount;
_a = JSII_RTTI_SYMBOL_1;
TerraformCount[_a] = { fqn: "cdktf.TerraformCount", version: "0.21.0" };
