// cdktf/lib/terraform-local.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraformLocal = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const terraform_element_1 = require("./terraform-element");
const tokens_1 = require("./tokens");
const tfExpression_1 = require("./tfExpression");
// eslint-disable-next-line jsdoc/require-jsdoc
class TerraformLocal extends terraform_element_1.TerraformElement {
    constructor(scope, id, expression) {
        super(scope, id, "local");
        this._expression = expression;
    }
    set expression(value) {
        this._expression = value;
    }
    get expression() {
        return tokens_1.Token.asAny(this.interpolation());
    }
    get asString() {
        return tokens_1.Token.asString(this.interpolation());
    }
    get asNumber() {
        return tokens_1.Token.asNumber(this.interpolation());
    }
    get asList() {
        return tokens_1.Token.asList(this.interpolation());
    }
    get asStringMap() {
        return tokens_1.Token.asStringMap(this.interpolation());
    }
    get asNumberMap() {
        return tokens_1.Token.asNumberMap(this.interpolation());
    }
    get asBooleanMap() {
        return tokens_1.Token.asBooleanMap(this.interpolation());
    }
    get asAnyMap() {
        return tokens_1.Token.asAnyMap(this.interpolation());
    }
    get asBoolean() {
        return this.interpolation();
    }
    interpolation() {
        return (0, tfExpression_1.ref)(`local.${this.friendlyUniqueId}`, this.cdktfStack);
    }
    toHclTerraform() {
        return {
            locals: {
                [this.friendlyUniqueId]: {
                    value: this._expression,
                    isBlock: false,
                    type: "any",
                },
            },
        };
    }
    toTerraform() {
        return {
            locals: {
                [this.friendlyUniqueId]: this._expression,
            },
        };
    }
    toMetadata() {
        if (!Object.keys(this.rawOverrides).length) {
            return {};
        }
        return {
            overrides: {
                local: Object.keys(this.rawOverrides),
            },
        };
    }
    /**
     * @returns a string token referencing the value of this local
     */
    toString() {
        return this.asString;
    }
}
exports.TerraformLocal = TerraformLocal;
_a = JSII_RTTI_SYMBOL_1;
TerraformLocal[_a] = { fqn: "cdktf.TerraformLocal", version: "0.21.0" };
