// @cdktf/provider-random/lib/integer/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Integer = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/random/3.7.2/docs/resources/integer random_integer}
*/
class Integer extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Integer resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Integer to import
    * @param importFromId The id of the existing Integer that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/random/3.7.2/docs/resources/integer#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Integer to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "random_integer", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/random/3.7.2/docs/resources/integer random_integer} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IntegerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'random_integer',
            terraformGeneratorMetadata: {
                providerName: 'random',
                providerVersion: '3.7.2',
                providerVersionConstraint: '~> 3.1'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._keepers = config.keepers;
        this._max = config.max;
        this._min = config.min;
        this._seed = config.seed;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get keepers() {
        return this.getStringMapAttribute('keepers');
    }
    set keepers(value) {
        this._keepers = value;
    }
    resetKeepers() {
        this._keepers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keepersInput() {
        return this._keepers;
    }
    get max() {
        return this.getNumberAttribute('max');
    }
    set max(value) {
        this._max = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxInput() {
        return this._max;
    }
    get min() {
        return this.getNumberAttribute('min');
    }
    set min(value) {
        this._min = value;
    }
    // Temporarily expose input value. Use with caution.
    get minInput() {
        return this._min;
    }
    // result - computed: true, optional: false, required: false
    get result() {
        return this.getNumberAttribute('result');
    }
    get seed() {
        return this.getStringAttribute('seed');
    }
    set seed(value) {
        this._seed = value;
    }
    resetSeed() {
        this._seed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get seedInput() {
        return this._seed;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            keepers: cdktf.hashMapper(cdktf.stringToTerraform)(this._keepers),
            max: cdktf.numberToTerraform(this._max),
            min: cdktf.numberToTerraform(this._min),
            seed: cdktf.stringToTerraform(this._seed),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            keepers: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._keepers),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            max: {
                value: cdktf.numberToHclTerraform(this._max),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            min: {
                value: cdktf.numberToHclTerraform(this._min),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            seed: {
                value: cdktf.stringToHclTerraform(this._seed),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Integer = Integer;
_a = JSII_RTTI_SYMBOL_1;
Integer[_a] = { fqn: "@cdktf/provider-random.integer.Integer", version: "12.0.0" };
// =================
// STATIC PROPERTIES
// =================
Integer.tfResourceType = "random_integer";
