// @cdktf/provider-random/lib/shuffle/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Shuffle = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/random/3.7.2/docs/resources/shuffle random_shuffle}
*/
class Shuffle extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Shuffle resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Shuffle to import
    * @param importFromId The id of the existing Shuffle that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/random/3.7.2/docs/resources/shuffle#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Shuffle to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "random_shuffle", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/random/3.7.2/docs/resources/shuffle random_shuffle} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ShuffleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'random_shuffle',
            terraformGeneratorMetadata: {
                providerName: 'random',
                providerVersion: '3.7.2',
                providerVersionConstraint: '~> 3.1'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._input = config.input;
        this._keepers = config.keepers;
        this._resultCount = config.resultCount;
        this._seed = config.seed;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get input() {
        return this.getListAttribute('input');
    }
    set input(value) {
        this._input = value;
    }
    // Temporarily expose input value. Use with caution.
    get inputInput() {
        return this._input;
    }
    get keepers() {
        return this.getStringMapAttribute('keepers');
    }
    set keepers(value) {
        this._keepers = value;
    }
    resetKeepers() {
        this._keepers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keepersInput() {
        return this._keepers;
    }
    // result - computed: true, optional: false, required: false
    get result() {
        return this.getListAttribute('result');
    }
    get resultCount() {
        return this.getNumberAttribute('result_count');
    }
    set resultCount(value) {
        this._resultCount = value;
    }
    resetResultCount() {
        this._resultCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resultCountInput() {
        return this._resultCount;
    }
    get seed() {
        return this.getStringAttribute('seed');
    }
    set seed(value) {
        this._seed = value;
    }
    resetSeed() {
        this._seed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get seedInput() {
        return this._seed;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            input: cdktf.listMapper(cdktf.stringToTerraform, false)(this._input),
            keepers: cdktf.hashMapper(cdktf.stringToTerraform)(this._keepers),
            result_count: cdktf.numberToTerraform(this._resultCount),
            seed: cdktf.stringToTerraform(this._seed),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            input: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._input),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            keepers: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._keepers),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            result_count: {
                value: cdktf.numberToHclTerraform(this._resultCount),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            seed: {
                value: cdktf.stringToHclTerraform(this._seed),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Shuffle = Shuffle;
_a = JSII_RTTI_SYMBOL_1;
Shuffle[_a] = { fqn: "@cdktf/provider-random.shuffle.Shuffle", version: "12.0.0" };
// =================
// STATIC PROPERTIES
// =================
Shuffle.tfResourceType = "random_shuffle";
