// cdktf/lib/backends/http-backend.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataTerraformRemoteStateHttp = exports.HttpBackend = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const terraform_backend_1 = require("../terraform-backend");
const util_1 = require("../util");
const terraform_remote_state_1 = require("../terraform-remote-state");
// eslint-disable-next-line jsdoc/require-jsdoc
class HttpBackend extends terraform_backend_1.TerraformBackend {
    constructor(scope, props) {
        super(scope, "backend", "http");
        this.props = props;
    }
    synthesizeAttributes() {
        return (0, util_1.keysToSnakeCase)({ ...this.props });
    }
    synthesizeHclAttributes() {
        return (0, util_1.keysToSnakeCase)({ ...this.props });
    }
    getRemoteStateDataSource(scope, name, _fromStack) {
        return new DataTerraformRemoteStateHttp(scope, name, this.props);
    }
}
exports.HttpBackend = HttpBackend;
_a = JSII_RTTI_SYMBOL_1;
HttpBackend[_a] = { fqn: "cdktf.HttpBackend", version: "0.21.0" };
// eslint-disable-next-line jsdoc/require-jsdoc
class DataTerraformRemoteStateHttp extends terraform_remote_state_1.TerraformRemoteState {
    constructor(scope, id, config) {
        super(scope, id, "http", config);
    }
}
exports.DataTerraformRemoteStateHttp = DataTerraformRemoteStateHttp;
_b = JSII_RTTI_SYMBOL_1;
DataTerraformRemoteStateHttp[_b] = { fqn: "cdktf.DataTerraformRemoteStateHttp", version: "0.21.0" };
