// @cdktf/provider-time/lib/sleep/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sleep = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/time/0.13.1/docs/resources/sleep time_sleep}
*/
class Sleep extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Sleep resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Sleep to import
    * @param importFromId The id of the existing Sleep that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/time/0.13.1/docs/resources/sleep#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Sleep to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "time_sleep", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/time/0.13.1/docs/resources/sleep time_sleep} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SleepConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'time_sleep',
            terraformGeneratorMetadata: {
                providerName: 'time',
                providerVersion: '0.13.1',
                providerVersionConstraint: '~> 0.7'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._createDuration = config.createDuration;
        this._destroyDuration = config.destroyDuration;
        this._triggers = config.triggers;
    }
    get createDuration() {
        return this.getStringAttribute('create_duration');
    }
    set createDuration(value) {
        this._createDuration = value;
    }
    resetCreateDuration() {
        this._createDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createDurationInput() {
        return this._createDuration;
    }
    get destroyDuration() {
        return this.getStringAttribute('destroy_duration');
    }
    set destroyDuration(value) {
        this._destroyDuration = value;
    }
    resetDestroyDuration() {
        this._destroyDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get destroyDurationInput() {
        return this._destroyDuration;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get triggers() {
        return this.getStringMapAttribute('triggers');
    }
    set triggers(value) {
        this._triggers = value;
    }
    resetTriggers() {
        this._triggers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get triggersInput() {
        return this._triggers;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            create_duration: cdktf.stringToTerraform(this._createDuration),
            destroy_duration: cdktf.stringToTerraform(this._destroyDuration),
            triggers: cdktf.hashMapper(cdktf.stringToTerraform)(this._triggers),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            create_duration: {
                value: cdktf.stringToHclTerraform(this._createDuration),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            destroy_duration: {
                value: cdktf.stringToHclTerraform(this._destroyDuration),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            triggers: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._triggers),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Sleep = Sleep;
_a = JSII_RTTI_SYMBOL_1;
Sleep[_a] = { fqn: "@cdktf/provider-time.sleep.Sleep", version: "11.0.0" };
// =================
// STATIC PROPERTIES
// =================
Sleep.tfResourceType = "time_sleep";
